internal/controller/nicconfigurationtemplate_controller.go

```diff
@@ -43,26 +57,254 @@ type NicConfigurationTemplateReconciler struct {
 //+kubebuilder:rbac:groups="",resources=pods,verbs=list
 //+kubebuilder:rbac:groups="",resources=pods/eviction,verbs=create;delete;get;list;patch;update;watch
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the NicConfigurationTemplate object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.3/pkg/reconcile
+// Reconcile reconciles the NicConfigurationTemplate object
 func (r *NicConfigurationTemplateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	if req.Name != nicConfigurationTemplateSyncEventName || req.Namespace != "" {
+		return reconcile.Result{}, nil
+	}
+	log.Log.Info("Reconciling NicConfigurationTemplates")
+
+	templateList := &v1alpha1.NicConfigurationTemplateList{}
+	err := r.List(ctx, templateList)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+	log.Log.V(2).Info(fmt.Sprintf("listed %d templates: %s", len(templateList.Items), templateList.Items))
+
+	deviceList := &v1alpha1.NicDeviceList{}
+	err = r.List(ctx, deviceList)
+	if err != nil {
+		log.Log.Error(err, "Failed to list NicDevices")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	log.Log.V(2).Info(fmt.Sprintf("listed %d devices: %s", len(deviceList.Items), deviceList.Items))
+
+	nodeList := &v1.NodeList{}
+	err = r.List(ctx, nodeList)
+	if err != nil {
+		log.Log.Error(err, "Failed to list cluster nodes")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	log.Log.V(2).Info(fmt.Sprintf("listed %d nodes: %s", len(nodeList.Items), nodeList.Items))
+
+	nodeMap := map[string]*v1.Node{}
+	for _, node := range nodeList.Items {
+		nodeMap[node.Name] = &node
+	}
+
+	for _, device := range deviceList.Items {
+		node, ok := nodeMap[device.Status.Node]
+		if !ok {
+			log.Log.Info("device doesn't match any node, skipping", "device", device.Name)
+			continue
+		}
+
+		var matchingTemplates []*v1alpha1.NicConfigurationTemplate
+
+		for _, template := range templateList.Items {
+			if !deviceMatchesSelectors(&device, &template, node) {
+				err := r.dropDeviceFromStatus(ctx, device.Name, &template)
```

Comment:
lets have a single API call to update status instead of many even if it means we do more work here.

Reply:
Done, now status updates once for each template